### ISSUE TITLE
Extract to the underscore identifier (Spec section 11.8.4) doesn't work

### ISSUE THREAD

**Issue by pofig37** [p4lang/p4c#774]
Compiling following test code fails:
```
#include <core.p4>

header Header {
    bit<32> data;
}

parser p0(packet_in p, out Header h) {
    state start {
        p.extract<Header>(_);
        transition next;
    }

    state next {
        p.extract(h);
        transition accept;
    }
}

parser proto(packet_in p, out Header h);
package top(proto p);
top(p0()) main;
```
fails with the  following error:
```
parser-extract-bug.p4(25):syntax error, unexpected _, expecting )
        p.extract<Header>(_
                          ^
error: 1 errors encountered, aborting compilation
```

**Comment by mbudiu-vmw** [comment]
This one is actually quite tricky. The problem is in the grammar; currently it does not accept _ as an expression. An attempt to add _ as a legal expression makes the grammar ambiguous: the compiler cannot tell whether _ is a type or an expression. @ChrisDodd : do you see a workaround for this?

In the worst case we may need to go back to the spec and disallow the use of the underscore identifier for expressions...

**Comment by jafingerhut** [comment]
At least for the case of an extract(_) method call, I believe the spec already mentions that another way to achieve the same effect is to use advance().  Note that regardless of which of these 2 methods is allowed or used, you could not later emit that header, so it is effectively removed from the incoming packet by doing these things in a parser.

In terms of what is possible semantically, there is no loss of functionality if _ is disallowed as an expression, is there?  It seems the same as declaring a temporary variable that can be put in place of the underscore, then never using that temporary variable's value anywhere.

**Comment by mbudiu-vmw** [comment]
Yes, we can still write the same set of programs.

**Comment by ChrisDodd** [comment]
Allowing `_` as an expression anywhere is hard to parse, but I'm not sure it makes sense.  Allowing it as an argument is simpler and doesn't introduce any conflicts.

    argument
        : expression  { $$ = $1; }
        | "_"         { $$ = new IR::DefaultExpression(@1); }
        ;

**Comment by mbudiu-vmw** [comment]
Thinking about this it only makes sense to allow it for out parameters. So we may need to amend the spec anyway.

**Comment by mbudiu-vmw** [comment]
This should work now; there is a pending fix to the spec too.
